presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/WeightedReservoirSampleStateStrategy.java

```diff
@@ -0,0 +1,140 @@
+        implements DifferentialEntropyStateStrategy
+{
+    private final WeightedDoubleReservoirSample reservoir;
+
+    public WeightedReservoirSampleStateStrategy(long size)
+    {
+        if (size <= 0) {
+            throw new PrestoException(
+                    INVALID_FUNCTION_ARGUMENT,
+                    format("Sample size %s must be positive", size));
+        }
+
+        reservoir = new WeightedDoubleReservoirSample((int) size);
+    }
+
+    private WeightedReservoirSampleStateStrategy(WeightedReservoirSampleStateStrategy other)
+    {
+        reservoir = other.reservoir.clone();
+    }
+
+    private WeightedReservoirSampleStateStrategy(WeightedDoubleReservoirSample reservoir)
+    {
+        this.reservoir = reservoir;
+    }
+
+    @Override
+    public void validateParameters(
+            long bucketCount,
+            double sample,
+            double weight,
+            double min,
+            double max)
+    {
+        throw new PrestoException(
+                INVALID_FUNCTION_ARGUMENT,
+                format("Unsupported parameters for %s", getClass().getSimpleName()));
+    }
+
+    @Override
+    public void validateParameters(
+            long size,
+            double sample,
+            double weight)
+    {
+        if (weight < 0.0) {
+            throw new PrestoException(
+                    INVALID_FUNCTION_ARGUMENT,
+                    format("Weight %s must be non-negative", weight));
+        }
+
+        if (size != reservoir.getMaxSamples()) {
+            throw new PrestoException(
+                    INVALID_FUNCTION_ARGUMENT,
+                    format("Inconsistent size %s with previous size %s", size, reservoir.getMaxSamples()));
+        }
+    }
+
+    @Override
+    public void validateParameters(long bucketCount, double sample)
+    {
+        throw new PrestoException(
+                INVALID_FUNCTION_ARGUMENT,
+                format("Unweighted operation should not be called for %s", getClass().getSimpleName()));
+    }
+
+    @Override
+    public void mergeWith(DifferentialEntropyStateStrategy other)
+    {
+        reservoir.mergeWith(((WeightedReservoirSampleStateStrategy) other).reservoir);
+    }
+
+    @Override
+    public void add(double value, double weight)
+    {
+        reservoir.add(value, weight);
+    }
+
+    @Override
+    public double calculateEntropy()
+    {
+        double[] samples = reservoir.stream().toArray();
+        return calculateEntropyFromSamples(samples);
+    }
+
+    @Override
+    public long getEstimatedSize()
+    {
+        return reservoir.estimatedInMemorySize();
+    }
+
+    @Override
+    public int getRequiredBytesForSerialization()
+    {
+        return reservoir.getRequiredBytesForSerialization();
+    }
+
+    public static WeightedReservoirSampleStateStrategy deserialize(SliceInput input)
+    {
+        WeightedDoubleReservoirSample reservoir = WeightedDoubleReservoirSample.deserialize(input);
```

Comment:
I'd inline this variable:

`return new WeightedReservoirSampleStateStrategy(WeightedDoubleReservoirSample.deserialize(input));`